pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -282,6 +282,10 @@ func ValidateServiceaccountMapping(p *v1beta1.PipelineSpec, pr *v1beta1.Pipeline
 		pipelineTasks[task.Name] = task.Name
 	}
 
+	for _, task := range p.Finally {
+		pipelineTasks[task.Name] = task.Name
+	}
+
 	for _, name := range pr.Spec.ServiceAccountNames {
```

Comment:
Interesting, this is a miss when we deprecated `ServiceAccountNames` :thinking: it shouldn't be looking here (or at least it should be sharing the same code as what the reconciler does)

Reply:
> > @pritidesai just looked at the code and it seems that even by using `taskRunSpecs` you would have the same issue. See [here](https://github.com/tektoncd/pipeline/blob/373081364f4bf2f1776e3ea7ed1edda2f71a53f0/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L266), it doesn't iterate over the finally tasks which means you won't be able to set a service account for a finally task and fail with the same issue as what this PR tries to fix.
> > Should I fix that part and include it in this PR?
> 
> Definitely yes wink

Done, also edited the PR description to match the current state.

Reply:
I am still trying to understand this but it looks a bit weird so far 😕 

Reconciler runs validation on `pipelineRun.Spec.ServiceAccountNames` and `pipelineRun.Spec.taskRunSpecs`, making sure the tasks specified in these sections exist under `tasks` and `finally`:

https://github.com/tektoncd/pipeline/blob/4fdd018d4b54b5cbd44cf03fd1450ef727942e3e/pkg/reconciler/pipelinerun/pipelinerun.go#L443

and

https://github.com/tektoncd/pipeline/blob/4fdd018d4b54b5cbd44cf03fd1450ef727942e3e/pkg/reconciler/pipelinerun/pipelinerun.go#L451

But, `resolvePipelineState` resolves task with `pr.Spec.ServiceAccountName`:

https://github.com/tektoncd/pipeline/blob/4fdd018d4b54b5cbd44cf03fd1450ef727942e3e/pkg/reconciler/pipelinerun/pipelinerun.go#L471

https://github.com/tektoncd/pipeline/blob/4fdd018d4b54b5cbd44cf03fd1450ef727942e3e/pkg/reconciler/pipelinerun/pipelinerun.go#L266

Only applies to `bundle` at this point, i.e. bundles does not honor `pr.Spec.ServiceAccountNames` or `pr.Spec.TaskRunSpecs` 🤔  

https://github.com/tektoncd/pipeline/blob/4fdd018d4b54b5cbd44cf03fd1450ef727942e3e/pkg/reconciler/taskrun/resources/taskref.go#L53

Looking further 🔍

Reply:
alright, `createTaskRun` resolves service account name in expected order: 

https://github.com/tektoncd/pipeline/blob/4fdd018d4b54b5cbd44cf03fd1450ef727942e3e/pkg/reconciler/pipelinerun/pipelinerun.go#L654

First, read serviceAccountName specified in `pr.Spec.ServiceAccountNames`:

https://github.com/tektoncd/pipeline/blob/4fdd018d4b54b5cbd44cf03fd1450ef727942e3e/pkg/apis/pipeline/v1beta1/pipelinerun_types.go#L141

which get replaced with `pr.Spec.TaskRunSpecs` if specified:

https://github.com/tektoncd/pipeline/blob/4fdd018d4b54b5cbd44cf03fd1450ef727942e3e/pkg/apis/pipeline/v1beta1/pipelinerun_types.go#L438-L439

Reply:
> it shouldn't be looking here `pr.Spec.ServiceAccountNames` (or at least it should be sharing the same code as what the reconciler does)

It's looking here to validate which is fine from the validation perspective, reconciler reads `serviceAccountName`, `serviceAccountNames`, and `taskRunSpecs`, and sets the service account name in expected order.